### PR TITLE
Fix false positive for unknown emits definition in `vue/require-explicit-emits` rule

### LINKED ISSUE #1748: `vue/require-explicit-emits`  doesn't support non-inline props declarations
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- ESLint version:8.5.0
- eslint-plugin-vue version:8.2.0
- Node version:16.13.0
- Operating System:win10

**Please show your full configuration:**

```javascript
module.exports = {
  extends: [
    'plugin:vue/vue3-recommended',
    'eslint:recommended',
    'plugin:import/recommended',
    'plugin:import/typescript',
    'plugin:@typescript-eslint/recommended',
    'plugin:prettier/recommended'
  ],

  parserOptions: {
    parser: '@typescript-eslint/parser',
    ecmaVersion: 2020,
    sourceType: 'module',
    extraFileExtensions: ['.vue']
  },

  plugins: ['@typescript-eslint'],

  settings: {
    'import/resolver': {
      typescript: {}
    },
    'import/ignore': ['node_modules', '\\.(coffee|scss|css|less|hbs|svg|json)$']
  },

  env: {
    es6: true,
    node: true,
    jest: true,
    browser: true
  },

  globals: {
    __VERSION__: true
  },

  rules: {
    // eslint-plugin-import
    'import/named': 'error',
    'import/default': 'off',
    'import/namespace': 'off',
    'import/export': 'error',
    'import/no-named-as-default': 'off',
    'import/no-named-as-default-member': 'error',
    'import/no-deprecated': 'off',

    // disallow require()
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-commonjs.md
    'import/no-commonjs': 'off',

    // disallow AMD require/define
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-amd.md
    'import/no-amd': 'error',

    // No Node.js builtin modules
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-nodejs-modules.md
    'import/no-nodejs-modules': 'off',

    // disallow non-import statements appearing before import statements
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/first.md
    // 'import/first': 'error',

    // disallow non-import statements appearing before import statements
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/imports-first.md
    // deprecated: use `import/first`
    'import/imports-first': 'off',

    // disallow duplicate imports
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-duplicates.md
    'import/no-duplicates': 'error',

    // Require a newline after the last import/require in a group
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/newline-after-import.md
    'import/newline-after-import': 'error',

    // Require modules with a single export to use a default export
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/prefer-default-export.md
    // 'import/prefer-default-export': 'error',

    // Restrict which files can be imported in a given folder
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-restricted-paths.md
    'import/no-restricted-paths': 'off',

    // Forbid modules to have too many dependencies
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/max-dependencies.md
    'import/max-dependencies': ['off', { max: 10 }],

    // Forbid import of modules using absolute paths
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-absolute-path.md
    'import/no-absolute-path': 'error',

    // Forbid require() calls with expressions
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-dynamic-require.md
    // 'import/no-dynamic-require': 'error',

    // prevent importing the submodules of other modules
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-internal-modules.md
    'import/no-internal-modules': [
      'off',
      {
        allow: []
      }
    ],

    // Warn if a module could be mistakenly parsed as a script by a consumer
    // leveraging Unambiguous JavaScript Grammar
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/unambiguous.md
    // this should not be enabled until this proposal has at least been *presented* to TC39.
    // At the moment, it's not a thing.
    'import/unambiguous': 'off',

    // Forbid Webpack loader syntax in imports
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-webpack-loader-syntax.md
    'import/no-webpack-loader-syntax': 'error',

    // Prevent unassigned imports
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unassigned-import.md
    // importing for side effects is perfectly acceptable, if you need side effects.
    'import/no-unassigned-import': 'off',

    // Prevent importing the default as if it were named
    // https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-named-default.md
    'import/no-named-default': 'error',

    // Reports if a module's default export is unnamed
    // https://github.com/benmosher/eslint-plugin-import/blob/d9b712ac7fd1fddc391f7b234827925c160d956f/docs/rules/no-anonymous-default-export.md
    'import/no-anonymous-default-export': [
      'off',
      {
        allowArray: false,
        allowArrowFunction: false,
        allowAnonymousClass: false,
        allowAnonymousFunction: false,
        allowLiteral: false,
        allowObject: false
      }
    ],

    // This rule enforces that all exports are declared at the bottom of the file.
    // https://github.com/benmosher/eslint-plugin-import/blob/98acd6afd04dcb6920b81330114e146dc8532ea4/docs/rules/exports-last.md
    'import/exports-last': 'off',

    // Reports when named exports are not grouped together in a single export declaration
    // or when multiple assignments to CommonJS module.exports or exports object are present
    // in a single file.
    // https://github.com/benmosher/eslint-plugin-import/blob/44a038c06487964394b1e15b64f3bd34e5d40cde/docs/rules/group-exports.md
    'import/group-exports': 'off',

    // forbid default exports. this is a terrible rule, do not use it.
    // https://github.com/benmosher/eslint-plugin-import/blob/44a038c06487964394b1e15b64f3bd34e5d40cde/docs/rules/no-default-export.md
    'import/no-default-export': 'off',

    // Prohibit named exports. this is a terrible rule, do not use it.
    // https://github.com/benmosher/eslint-plugin-import/blob/1ec80fa35fa1819e2d35a70e68fb6a149fb57c5e/docs/rules/no-named-export.md
    'import/no-named-export': 'off',

    // Forbid a module from importing itself
    // https://github.com/benmosher/eslint-plugin-import/blob/44a038c06487964394b1e15b64f3bd34e5d40cde/docs/rules/no-self-import.md
    'import/no-self-import': 'error',

    // Forbid cyclical dependencies between modules
    // https://github.com/benmosher/eslint-plugin-import/blob/d81f48a2506182738409805f5272eff4d77c9348/docs/rules/no-cycle.md
    'import/no-cycle': ['error', { maxDepth: '∞' }],

    // Ensures that there are no useless path segments
    // https://github.com/benmosher/eslint-plugin-import/blob/ebafcbf59ec9f653b2ac2a0156ca3bcba0a7cf57/docs/rules/no-useless-path-segments.md
    'import/no-useless-path-segments': ['error', { commonjs: true }],

    // dynamic imports require a leading comment with a webpackChunkName
    // https://github.com/benmosher/eslint-plugin-import/blob/ebafcbf59ec9f653b2ac2a0156ca3bcba0a7cf57/docs/rules/dynamic-import-chunkname.md
    'import/dynamic-import-chunkname': [
      'off',
      {
        importFunctions: [],
        webpackChunknameFormat: '[0-9a-zA-Z-_/.]+'
      }
    ],

    // Use this rule to prevent imports to folders in relative parent paths.
    // https://github.com/benmosher/eslint-plugin-import/blob/c34f14f67f077acd5a61b3da9c0b0de298d20059/docs/rules/no-relative-parent-imports.md
    'import/no-relative-parent-imports': 'off',

    // Reports modules without any exports, or with unused exports
    // https://github.com/benmosher/eslint-plugin-import/blob/f63dd261809de6883b13b6b5b960e6d7f42a7813/docs/rules/no-unused-modules.md
    'import/no-unused-modules': [
      'off',
      {
        ignoreExports: [],
        missingExports: true,
        unusedExports: true
      }
    ],

    'import/order': 'off',
    'import/extensions': 'off',
    'import/no-unresolved': 'off',
    'import/prefer-default-export': 'off',
    'import/no-extraneous-dependencies': 'off',
    // eslint-plugin-vue
    'vue/no-v-html': 'off',
    'vue/attributes-order': 'off',
    'vue/require-v-for-key': 'off',
    'vue/require-default-prop': 'off',
    'vue/no-unused-components': 'off',
    'vue/return-in-computed-property': 'off',
    'vue/multi-word-component-names': 'off',
    // typescript-eslint
    '@typescript-eslint/no-empty-interface': 'off',
    '@typescript-eslint/no-empty-function': 'off',
    '@typescript-eslint/no-var-requires': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/ban-types': 'off',
    '@typescript-eslint/ban-ts-comment': 'off',
    '@typescript-eslint/no-non-null-assertion': 'off'
  },

  overrides: [
    {
      files: ['*.vue'],
      parser: require.resolve('vue-eslint-parser')
    }

  ]
}

```

**What did you do?**	

```typerscript
import type { ExtractPropTypes } from 'vue'
import { createNamespace } from '@ly-element/utils/create'

const [name] = createNamespace('button')

const buttonProps = {
  ripple: {
    type: Boolean,
    default: false
  }
}
const buttonEmits = {
  click: (evt: MouseEvent) => evt instanceof MouseEvent
}

type ButtonProps = ExtractPropTypes<typeof buttonProps>
type ButtonEmits = ExtractPropTypes<typeof buttonEmits>

export { name, buttonEmits, buttonProps }
export type { ButtonProps, ButtonEmits }

```
```vue
<template>
  <button @click="add">rpple: {{ ripple }}</button>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { name, buttonProps, buttonEmits } from './base'

export default defineComponent({
  name,
  props: buttonProps,
  emits: buttonEmits,
  setup(props, { emit }) {
    const add = (event: MouseEvent) => {
      // eslint-disable-next-line vue/require-explicit-emits
      emit('click', event)
    }
    return {
      add
    }
  }
})
</script>

```

**What did you expect to happen?**

```vue
 emit('click', event) // There is no problem， but it actually reported an error
```

**What actually happened?**

```vue
 emit('click', event) // ESLint: The "click" event has been triggered but not declared on `emits` option.(vue/require-explicit-emits)
```
**Repository to reproduce this issue**

### PR DESCRIPTION
**Pull request by ota-meshi**
fixes #1748

### CODE REVIEW

File: lib/utils/index.js [hunk 1]
```diff
@@ -811,57 +815,18 @@ module.exports = {
...
-        )
-      }
-    )
-
-    if (!propsNode) {
-      return []
-    }
-
-    return getComponentPropsFromDefine(propsNode.value)
+    return getComponentPropsFromOptions(componentObject)
```
Comment (FloEdelmann) [inline lib/utils/index.js:821]: This is the only statement in the `getComponentProps` function now. Can we rename it to `getComponentPropsFromOptions` (also everywhere it is used) and only export a reference to `getComponentPropsFromOptions`?

```js
module.exports = {
  // …
  getComponentPropsFromOptions,
  // …
}

function getComponentPropsFromOptions(…) { … }
```

Or rename `getComponentPropsFromOptions` to `getComponentProps` and export only a reference to it?

```js
module.exports = {
  // …
  getComponentProps,
  // …
}

function getComponentProps(…) { … }
```

Or don't rename anything, but export still only export a reference to it?

```js
module.exports = {
  // …
  getComponentProps: getComponentPropsFromOptions,
  // …
}

function getComponentPropsFromOptions(…) { … }
```

File: lib/utils/index.js [hunk 2]
```diff
@@ -811,57 +815,18 @@ module.exports = {
...
-        )
-      }
-    )
-
-    if (!emitsNode) {
-      return []
-    }
-
-    return getComponentEmitsFromDefine(emitsNode.value)
+    return getComponentEmitsFromOptions(componentObject)
```
Comment (FloEdelmann) [inline lib/utils/index.js:829]: Same thing here.

**Review by FloEdelmann — COMMENTED** [review]
Looks good to me. I haven't looked thoroughly at the logic changes in `utils` though. I noticed two minor code simplifications, but they are not merge-blocking.

File: lib/utils/index.js (hunk 1 above)
Comment (ota-meshi) [reply lib/utils/index.js]: Thank you for your review! I changed method name and to use the shorthand property.
However, for backward compatibility, it also defines a method with the old name.

**Review by FloEdelmann — APPROVED** [review]
Thanks for implementing the feedback!

**Pull request merged**